### ISSUE TITLE
Buffer limit API for message events, links, and annotations?

### ISSUE THREAD

**Issue by jgracenin** [census-instrumentation/opencensus-specs#138]
I have a Go service with a long running stream that had a memory leak due to unbounded storage of 
message events. This was a known limitation: census-instrumentation/opencensus-go#401 

gRPC's MAX_CONNECTION_AGE would help, but I also made a quick fork with a limit on message events, annotations, and links: census-instrumentation/opencensus-go#824 .

Do any other language implementations already limit these internal structures yet (events, links annotations) yet?

The PR also sparked a larger discussion on more granular limits for some future features (e.g. classified events, links annotations), but that's a larger API change and is probably better to be discussed here.

**Comment by songy23** [comment]
> Do any other language implementations already limit these internal structures yet (events, links annotations) yet?

Yes, in Java we have some [default limit for them](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/trace/config/TraceParams.java#L40-L43):
```java
  private static final int DEFAULT_SPAN_MAX_NUM_ATTRIBUTES = 32;
  private static final int DEFAULT_SPAN_MAX_NUM_ANNOTATIONS = 32;
  private static final int DEFAULT_SPAN_MAX_NUM_MESSAGE_EVENTS = 128;
  private static final int DEFAULT_SPAN_MAX_NUM_LINKS = 128;
```

Users can override these numbers with [TraceConfig.updateActiveTraceParams](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/trace/config/TraceConfig.java#L42).

**Comment by rakyll** [comment]
@songy23, can we update https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/TraceConfig.md to represent this? So, other languages can refer to the spec when implementing.

**Comment by songy23** [comment]
Sure, working on a PR now.

**Comment by g-easy** [comment]
https://github.com/census-instrumentation/opencensus-specs/issues/138#issuecomment-404962790 FYI C++ has the same.
